src/event.rs

```diff
@@ -77,6 +77,17 @@ pub enum Event<'a, T: 'static> {
     Suspended,
 
     /// Emitted when the application has been resumed.
+    ///
+    /// For consistency all platforms emit a `Resumed` event even if they don't themselves have a
+    /// formal suspend/resume lifecycle. For systems without a standard suspend/resume lifecycle
+    /// the `Resumed` event is is always emitted after the `NewEvents(StartCause::Init)` event.
+    ///
+    /// # Portability
+    ///
+    /// It's recommended that applications should only initialized their graphics context and create
+    /// a window after they have received their first `Resumed` event. Some systems
+    /// (specifically Android) won't allow applications to create a render surface until they are
+    /// resumed (on Android this also signals that the application has an associated Surface View).
     Resumed,
```

Comment:
Should the docs here also link back to `` [`Self::Suspended`] `` and state that it must be handled in terms of cleaning up the "render surface" again? And repeat that on the `Suspended` docs itself?

Reply:
> Should the docs here also link back to `` [`Self::Suspended`] `` and state that it must be handled in terms of cleaning up the "render surface" again? And repeat that on the `Suspended` docs itself?

To some extent I was thinking of trying to keep some of the Android specific discussion separate.

Although my initial motivation came from wanting to write code that runs on Android and on desktop which this helps with, the change itself isn't really about Android. It's really just a general consistency improvement that means applications can start relying on the Resumed event across all platforms - which apps can take or leave.

That said though I did specifically add the `# Portability` notes, and this is intended to be a small stepping stone towards making the winit API guide towards being portable by default so maybe we should really flesh out the discussion of Android details here. I'm just thinking it could be good to decouple that stuff in case we start iterating on docs that depend on other nuanced details that are Android specific and also potentially subject to change in the run up to 0.27.

In terms of providing a general recommendation that avoids being too Android specific it's tricky to know what to advise right now. On Android even, technically, there's a good chance your app won't blow up if you ignore `Suspended` events and just make sure to create a new surface on the next resume (there won't be any winit redraw events while suspended which is the most likely place where things will go wrong)

Reply:
> To some extent I was thinking of trying to keep some of the Android specific discussion separate.

Oh no, I didn't mention android anywhere :sweat_smile:. If you're generalizing `Resumed` it would be stupid to not generalize `Suspended` too, right? The user doesn't "need to know" that other platforms never invalidate their surfaces.

_However_.

Perhaps this _is_ relevant when a `window` is destroyed, for which a [separate event exists](https://docs.rs/winit/latest/winit/event/enum.WindowEvent.html#variant.Destroyed) :vomiting_face:?

_Piggy-backing on my rename request_:

I think I suggested `WindowEvent::Created` at some point, which could nicely complement this already-existing `WindowEvent::Destroyed`? Would that work out for Android and do away with `Suspended`/`Resumed` completely?

---

> On Android even, technically, there's a good chance your app won't blow up if you ignore `Suspended` events and just make sure to create a new surface on the next resume (there won't be any winit redraw events while suspended which is the most likely place where things will go wrong)

I recently implemented proper `Clone` + `Drop` ownership semantics for `NativeWindow` (which is _just_ a `Surface`, i.e. a producer) and I don't remember whether it's invalid to push contents into it when there's no consumer (i.e. no display).

Reply:
yeah, I took it as being android specific because it's the only platform where it could be prudent to destroy any rendering surfaces you have.

I'd prefer to keep any potential reworking of suspend as a separate topic/issue if possible for a number of reasons:

1. We're solving a known portability problem by consistently emitting `Resumed` events - and the same can't really be said for `Suspended`. (apps running on Android may want to destroy render surfaces but it's ok to not do that on other platforms which is different to the `Resumed` situation where we have mutually exclusive paths for initialzing Android vs other platforms)
2. One reasonable interpretation of the resulting behaviour is that most platforms will never actually "suspend" and so it remains reasonable that there is no `Suspended` event, except on mobile platforms that have a well defined lifecycle. Conceptually it's harder to argue that these other platforms don't "resume" though, so I don't see the slight asymmetry as a major concern right now (not in comparison to the benefit of at least making `Resumed` consistent).
3. For non-mobile platforms it may be technically impossible to guarantee that a `Suspended` event will be delivered because the operating doesn't provide the required lifecycle hooks that exist on Android / iOS (e.g. processes can simply be killed via signals)

Reply:
This discussion is a good example for highlighting why the Android NDK terminology is really unhelpful unfortunately...

What the Android NDK calls a `NativeWindow` is a `SurfaceView` in Java - and even that isn't the clearest name for explaining what the object represents.

Conceptually a SurfaceView on Android is somewhat comparable to a rendering API surface, such as an EGLSurface. A SurfaceView encapsulates the swap chain for rendering to the window (or rather for getting buffers that can be given to Surface Flinger). Android technically even has a Window API in Java which could even be considered the true "native window" but since you don't have to explicitly create a SurfaceView for a Window then it's not a distinction that ever comes up in practice.

The main difference between Android and other OSs, is that other OSs let you create your render surfaces (and allocate/configure a swap chain) manually for a given native window but on Android they effectively create the render surface for you. When you create an EGLSurface or similar on Android there is a bit of a white lie involved because you don't really own the backing of the surface like you would on other OSs. That's the crux of why Android is awkward here you have to be prepared for your EGLSurface (or equivalent) to become invalid. If comparing with other OSs then it's _not_ equivalent to the native window being destroyed - so terminology is important here.

Some consequences of this to consider are:
- Just because the Android NDK tells us that our 'window' has been destroyed / created doesn't mean that we should use the same terminology within the public API of Winit. In particular; if we want to expose the Android NDK "window created" and "window destroyed" events via Winit I would strongly recommend avoiding the word "window" and probably advocate for events like `WindowEvent::SurfaceCreated` `WindowEvent::SurfaceDestroyed` to avoid a lot of confusion with how they related to `WindowEvent::Destroyed`.
- Since Android's `NativeWindow` is actually just a rendering surface then certain questions about a "window" should be answerable without a surface - such as querying width/height which doesn't necessarily have to be done via the SurfaceView (conceptually the same as being able to query the width/height of an X11 window or wl_surface without necessarily looking at the width/height of an EGLSurface)

Reply:
> * and probably advocate for events like `WindowEvent::SurfaceCreated` `WindowEvent::SurfaceDestroyed` to avoid a lot of confusion with how they related to `WindowEvent::Destroyed`.

:+1: Yes, I guess it is better to add separate events for this then. The window (activity, really) technically stays alive but may pause because the user stops interacting with it, and separately from those lifecycle events the backing buffer of our `Surface` _producer_ may become invalid/recreated.

Then later we may map activity destruction to this (cc some thoughts about recently running into multiple `NativeActivity` _instances_ within the same executable process making the whole Rust Android stack fall apart).

Reply:
> > * and probably advocate for events like `WindowEvent::SurfaceCreated` `WindowEvent::SurfaceDestroyed` to avoid a lot of confusion with how they related to `WindowEvent::Destroyed`.
> 
> 👍 Yes, I guess it is better to add separate events for this then. The window (activity, really) technically stays alive but may pause because the user stops interacting with it, and separately from those lifecycle events the backing buffer of our `Surface` _producer_ may become invalid/recreated.
> 
> Then later we may map activity destruction to this (cc some thoughts about recently running into multiple `NativeActivity` _instances_ within the same executable process making the whole Rust Android stack fall apart).

Worried that we're spawning a lot of side discussions here but yeah, this was something I was unsure about recently with the activity/glue crates I was experimenting with. Actually my first pass would have supported multiple activities but I ended up trying to emulate the global getter API from ndk_glue to made it easier to port some existing code, but I think it would be relatively easy to re-instate support for multiple activities and also enable support for this in Winit too.

The main requirements are:
1. remove any global getters for android app/activity state
2. the main() entrypoint on Android needs to accept an application pointer
3. The application pointer needs to be given to Winit when creating an event loop
4. Downstream winit users need to be able to query the application from the event loop (similar to getting the X11/wayland display)

when I say "application pointer" I essentially mean an opaque struct that encapsulates the activity. E.g. in my `native-activity` glue crate that would correspond to this [`AndroidApp`](https://github.com/rib/agdk-rust/blob/864d782c241aa7e403786774ebf057ec134f93be/native-activity/src/lib.rs#L284) struct / API.

The main tricky thing is then with defining the C ABI entry point (`android_main()`) which wouldn't be quite as ergonomic without the global static API to get the `AndroidApp` since it would have to be passed as a `[transparent]` pointer - but still that might not be too bad.

Reply:
Yeah we're spawning a side-discussion here. I was doing the same with https://github.com/rust-windowing/android-ndk-rs/pull/223#issuecomment-1039225349 but that branch never reached main - in short because it requires your point 3. i.e. `The application pointer needs to be given to Winit when creating an event loop` which I think is a fair requirement to prevent a lot of mishaps _at compile-time_.

Though my idea was to actually pass `ndk::NativeActivity` around - it's an opaque pointer internally but if your `android_main()` is wrapped by `ndk_glue 0.6` and winit uses `ndk_glue 0.5`. We'd have to discuss this further in light of making `winit` support multiple Android backends concurrently (as well how _and where_ to design and maintain a consistent API, and where to place all the crates - `android-ndk-rs` has my preference of course).

IIRC I also "got stuck" at storing the rest of our Rust state in its `instance` `void *` userdata, but recently found an ancient comment in the `ndk` that it was used by a previous `android_native_app_glue`-something-something crate as well - we just have to make sure that the user can't touch it if we do.

Reply:
Yeah, I'd be interested to pick up a discussion about building on what I found with the `native-activity` and `game-activity` crates.

Those two crates demonstrate that it's possible to define a compatible API that can support both NativeActivity and GameActivity (and theoretically future activities such as a hypothetical `RustActivity`) and I think it would be relatively simple to tweak and remove the static global `AndroidApp` getter to also be able to support multiple activities. I think I even added winit API to pass an `AndroidApp` when creating an event loop in an iteration of that work.

The next step I had roughly been imagining was defining an `android-activity` crate that would effectively be a single point of configuration for what specific Activity your app is uses and it would basically be like a MUX for the real implementation (such as `native-activity` or `game-activity` or other-activity...

Maybe this could be done under `android-ndk-rs` somewhere, or under `rust-windowing`  - though I'd probably see it as something separate from `ndk-glue` at this point (though we could potentially re-purpose `ndk-glue` as that kind of MUX crate.